File: worker/test/protocol/messages.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { expect, test } from "vitest";

import { canonicalize } from "../../src/protocol/canonical";
import {
  admissionAuthorityKey,
  assertHeartbeatBudget,
  localLeaseDeadlineMs,
  parseHeartbeatRequest,
  parseHeartbeatResponse,
  parseLease,
  parseSessionRequest,
  parseSessionResponse,
} from "../../src/protocol/messages";

const fixtureRoot = join(
  dirname(fileURLToPath(import.meta.url)),
  "../../../tests/fixtures/protocol/v1",
);

const digest = "a".repeat(64);
const session = "b".repeat(64);

const validLease = {
  archivedDisabledAliases: ["repo-a"],
  canaryScaleSet: "canary-set",
  durationMs: 8000,
  expiry: "2026-01-01T00:00:09.000Z",
  fleetId: "example-fleet",
  holder: "portable",
  leaseGeneration: 3,
  localPolicyEpoch: 9,
  maxCapacity: 1,
  mode: "canary-only",
  policyDigest: digest,
  protocolVersion: 1,
  repositoryPolicyRevision: 4,
  serverEpoch: 2,
  sessionId: session,
};

const validHeartbeat = {
  epoch: 2,
  fenceGeneration: 1,
  fleetId: "example-fleet",
  holder: "portable",
  protocolVersion: 1,
  sequence: 1,
  sessionId: session,
  snapshot: {
    acquisitionMode: "enabled",
    assignedJobs: 2,
    buildId: digest,
    capacity: {
      available: 1,
      configured: 2,
      effective: 2,
      occupied: 1,
      queued: 0,
    },
    degraded: false,
    fleetAlias: "example-fleet",
    hostProfileId: "strict-linux-v1",
    lastTerminalAt: "2026-01-01T00:00:00.000Z",
    observedAt: "2026-01-01T00:00:01.000Z",
    oldestLiveAssignmentAgeMs: 1000,
    policyDigest: digest,
    policyEpoch: 1,
    repositoryPolicyRevision: 1,
    runningJobs: 1,
    unassignedReleasedListeners: 0,
  },
  timestamp: "2026-01-01T00:00:01.000Z",
};

const validHeartbeatResponse = {
  fleetId: "example-fleet",
  lease: validLease,
  maintenance: {
    kind: "none",
    leaseGeneration: validLease.leaseGeneration,
    sessionId: session,
  },
  noLeaseReason: null,
  protocolVersion: 1,
  receiptTime: "2026-01-01T00:00:01.000Z",
  routingState: "PORTABLE_CANARY",
  sequence: 1,
  sessionId: session,
};

test("session request rejects unknown, missing, and noncanonical fields", () => {
  const body = canonicalize({
    buildId: digest,
    fleetId: "example-fleet",
    nonce: session,
    protocolVersion: 1,
    timestamp: "2026-01-01T00:00:00.000Z",
  });
  expect(parseSessionRequest(body).fleetId).toBe("example-fleet");
  expect(() =>
    parseSessionRequest(body.replace("example-fleet", "Example")),
  ).toThrow();
  expect(() => parseSessionRequest('{"fleetId":"example-fleet"}')).toThrow();
  expect(() =>
    parseSessionRequest(
      canonicalize({
        buildId: digest,
        extra: true,
        fleetId: "example-fleet",
        nonce: session,
        protocolVersion: 1,
        timestamp: "2026-01-01T00:00:00.000Z",
      }),
    ),
  ).toThrow();
});

test("session response binds a fresh initial session and monotonic boundary", () => {
  const response = {
    epoch: 1,
    fleetId: "example-fleet",
    leaseGeneration: 1,
    leaseNotBefore: "2026-01-01T00:00:00.000Z",
    nonce: "a".repeat(64),
    protocolVersion: 1,
    receiptTime: "2026-01-01T00:00:00.000Z",
    sequence: 0,
    sessionId: "c".repeat(64),
  };
  expect(parseSessionResponse(canonicalize(response)).sequence).toBe(0);
  expect(() =>
    parseSessionResponse(canonicalize({ ...response, sequence: 1 })),
  ).toThrow();
  expect(() =>
    parseSessionResponse(
      canonicalize({ ...response, sessionId: response.nonce }),
    ),
  ).toThrow();
  expect(() =>
    parseSessionResponse(
      canonicalize({
        ...response,
        leaseNotBefore: "2025-12-31T23:59:59.999Z",
      }),
    ),
  ).toThrow();
});

test("Worker canonical messages match every frozen Go exchange fixture", () => {
  const fixtures = [
    ["session-request", parseSessionRequest],
    ["session-response", parseSessionResponse],
    ["heartbeat-request", parseHeartbeatRequest],
    ["heartbeat-response-lease", parseHeartbeatResponse],
    ["heartbeat-response-no-lease", parseHeartbeatResponse],
  ] as const;
  for (const [name, parse] of fixtures) {
    const document = JSON.parse(
      readFileSync(join(fixtureRoot, `${name}.json`), "utf8"),
    ) as unknown;
    const canonical = readFileSync(
      join(fixtureRoot, `${name}.canonical.txt`),
      "utf8",
    ).trimEnd();
    expect(canonicalize(document), name).toBe(canonical);
    expect(() => parse(canonical), name).not.toThrow();
  }
});

test("protocol scalars reject calendar-invalid timestamps and unsafe integers", () => {
  expect(() =>
    parseSessionRequest(
      canonicalize({
        buildId: digest,
        fleetId: "example-fleet",
        nonce: session,
        protocolVersion: 1,
        timestamp: "2026-02-31T00:00:00.000Z",
      }),
    ),
  ).toThrow();
  expect(() =>
    parseLease({
      ...validLease,
      serverEpoch: Number.MAX_SAFE_INTEGER + 1,
    }),
  ).toThrow();
});

test("heartbeat snapshot rejects internally inconsistent health", () => {
  expect(parseHeartbeatRequest(canonicalize(validHeartbeat)).sequence).toBe(1);
  expect(() =>
    parseHeartbeatRequest(
      canonicalize({
        ...validHeartbeat,
        snapshot: {
          ...validHeartbeat.snapshot,
          capacity: {
            ...validHeartbeat.snapshot.capacity,
            available: 2,
          },
        },
      }),
    ),
  ).toThrow();
  expect(() =>
    parseHeartbeatRequest(
      canonicalize({
        ...validHeartbeat,
        snapshot: {
          ...validHeartbeat.snapshot,
          assignedJobs: 0,
          runningJobs: 1,
        },
      }),
    ),
  ).toThrow();
  expect(() =>
    parseHeartbeatRequest(
      canonicalize({
        ...validHeartbeat,
        snapshot: {
          ...validHeartbeat.snapshot,
          lastTerminalAt: "2026-01-01T00:00:02.000Z",
        },
      }),
    ),
  ).toThrow();
});

test("heartbeat response is exact and binds lease state", () => {
  expect(
    parseHeartbeatResponse(canonicalize(validHeartbeatResponse)).lease?.mode,
  ).toBe("canary-only");
  expect(() =>
    parseHeartbeatResponse(
      canonicalize({
        ...validHeartbeatResponse,
        maintenance: { ...validHeartbeatResponse.maintenance, extra: true },
      }),
    ),
  ).toThrow();
  expect(() =>
    parseHeartbeatResponse(
      canonicalize({
        ...validHeartbeatResponse,
        noLeaseReason: "lease-disabled",
      }),
    ),
  ).toThrow();
  expect(() =>
    parseHeartbeatResponse(
      canonicalize({
        ...validHeartbeatResponse,
        sessionId: "d".repeat(64),
      }),
    ),
  ).toThrow();
});

test("heartbeat lease expiry exactly equals receipt plus bounded duration", () => {
  for (const lease of [
    { ...validLease, expiry: "2026-01-01T00:00:08.999Z" },
    { ...validLease, expiry: "2026-01-01T00:00:09.001Z" },
    { ...validLease, durationMs: 9_223_372_036_855 },
  ]) {
    expect(() =>
      parseHeartbeatResponse(
        canonicalize({ ...validHeartbeatResponse, lease }),
      ),
    ).toThrow();
  }
});

test("heartbeat response enforces routing holder and mode matrix", () => {
  const fullLease = {
    ...validLease,
    canaryScaleSet: null,
    maxCapacity: 2,
    mode: "enabled",
  };
  const response = {
    ...validHeartbeatResponse,
    lease: fullLease,
    maintenance: {
      ...validHeartbeatResponse.maintenance,
      leaseGeneration: fullLease.leaseGeneration,
    },
    routingState: "PORTABLE",
  };
  expect(parseHeartbeatResponse(canonicalize(response)).lease?.mode).toBe(
    "enabled",
  );
  for (const next of [
    { ...response, routingState: "HOSTED" },
    { ...response, routingState: "DRAINING_TO_HOSTED" },
    { ...response, routingState: "PORTABLE_CANARY" },
    { ...response, lease: { ...fullLease, holder: "legacy" } },
    {
      ...response,
      lease: { ...fullLease, canaryScaleSet: "canary-set" },
    },
    { ...response, routingState: "LEGACY" },
  ]) {
    expect(() => parseHeartbeatResponse(canonicalize(next))).toThrow();
  }
});

test("heartbeat response and standalone lease require positive generation", () => {
  const response = {
    ...validHeartbeatResponse,
    lease: { ...validLease, leaseGeneration: 0 },
    maintenance: {
      ...validHeartbeatResponse.maintenance,
      leaseGeneration: 0,
    },
  };
  expect(() => parseHeartbeatResponse(canonicalize(response))).toThrow();
  expect(() =>
    parseHeartbeatResponse(
      canonicalize({
        ...response,
        lease: null,
        noLeaseReason: "routing-hosted",
        routingState: "HOSTED",
      }),
    ),
  ).toThrow();
  expect(() => parseLease({ ...validLease, leaseGeneration: 0 })).toThrow();
});

test("lease validation rejects unsorted aliases and wrong canary shape", () => {
  const lease = parseLease(validLease);
  expect(lease.archivedDisabledAliases).toEqual(["repo-a"]);
  expect(() =>
    parseLease({
      ...validLease,
      archivedDisabledAliases: ["repo-b", "repo-a"],
    }),
  ).toThrow();
  expect(() =>
    parseLease({
      ...validLease,
      archivedDisabledAliases: ["repo-a", "repo-a"],
    }),
  ).toThrow();
  expect(() => parseLease({ ...validLease, extra: 1 })).toThrow();
  expect(() =>
    parseLease({ ...validLease, mode: "enabled", maxCapacity: 0 }),
  ).toThrow();
  expect(() => parseLease({ ...validLease, canaryScaleSet: null })).toThrow();
});

test("admission authority key ignores renewal envelope fields", () => {
  const left = parseLease(validLease);
  const right = parseLease({
    ...validLease,
    expiry: "2026-01-01T00:00:10.000Z",
  });
  expect(admissionAuthorityKey(left)).toBe(admissionAuthorityKey(right));
  const changed = parseLease({
    ...validLease,
    durationMs: 9000,
    expiry: "2026-01-01T00:00:10.000Z",
  });
  expect(admissionAuthorityKey(left)).not.toBe(admissionAuthorityKey(changed));
});

test("heartbeat inequality requires every symbolic term", () => {
  assertHeartbeatBudget({
    leaseDurationMs: 20,
    maxAttemptIntervalMs: 4,
    deadlineMs: 3,
    shorteningMarginMs: 2,
    lostRenewals: 1,
  });
  expect(() =>
    assertHeartbeatBudget({
      leaseDurationMs: 10,
      maxAttemptIntervalMs: 4,
      deadlineMs: 3,
      shorteningMarginMs: 2,
      lostRenewals: 1,
    }),
  ).toThrow();
  expect(() =>
    assertHeartbeatBudget({
      leaseDurationMs: 20,
      maxAttemptIntervalMs: 4,
      deadlineMs: 3,
      shorteningMarginMs: 2,
      lostRenewals: 0,
    }),
  ).toThrow();
  expect(() =>
    assertHeartbeatBudget({
      leaseDurationMs: 20.5,
      maxAttemptIntervalMs: 4,
      deadlineMs: 3,
      shorteningMarginMs: 2,
      lostRenewals: 1,
    }),
  ).toThrow();
  expect(() =>
    assertHeartbeatBudget({
      leaseDurationMs: 20,
      maxAttemptIntervalMs: 4,
      deadlineMs: 3,
      shorteningMarginMs: 2,
      lostRenewals: 1.5,
    }),
  ).toThrow();
});

test("send-anchored local deadline is strictly shorter than server duration", () => {
  expect(localLeaseDeadlineMs(1_000, 8_000, 1_000)).toBe(8_000);
  expect(() => localLeaseDeadlineMs(1_000, 8_000, 8_000)).toThrow();
  expect(() => localLeaseDeadlineMs(1_000, 8_000, 0)).toThrow();
});

File: internal/failoverclient/messages_test.go
package failoverclient

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

func readProtocolFixture(t *testing.T, name string) []byte {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(fixtureRoot(t), name))
	if err != nil {
		t.Fatalf("read %s: %v", name, err)
	}
	return bytesTrimNL(raw)
}

func mutateProtocolFixture(
	t *testing.T,
	name string,
	mutate func(map[string]any),
) []byte {
	t.Helper()
	var document map[string]any
	if err := json.Unmarshal(readProtocolFixture(t, name), &document); err != nil {
		t.Fatalf("unmarshal %s: %v", name, err)
	}
	mutate(document)
	raw, err := CanonicalJSON(document)
	if err != nil {
		t.Fatalf("canonical %s mutation: %v", name, err)
	}
	return raw
}

func mutateLeaseFixture(t *testing.T, mutate func(map[string]any)) []byte {
	t.Helper()
	response := mutateProtocolFixture(t, "heartbeat-response-lease.canonical.txt", func(response map[string]any) {
		mutate(response["lease"].(map[string]any))
	})
	var document map[string]json.RawMessage
	if err := json.Unmarshal(response, &document); err != nil {
		t.Fatalf("unmarshal lease response: %v", err)
	}
	return document["lease"]
}

func TestParseFrozenProtocolMessages(t *testing.T) {
	sessionRequest, err := ParseSessionRequestV1(readProtocolFixture(t, "session-request.canonical.txt"))
	if err != nil {
		t.Fatalf("ParseSessionRequestV1: %v", err)
	}
	if sessionRequest.Nonce != "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" {
		t.Fatalf("session nonce = %q", sessionRequest.Nonce)
	}

	sessionResponse, err := ParseSessionResponseV1(readProtocolFixture(t, "session-response.canonical.txt"))
	if err != nil {
		t.Fatalf("ParseSessionResponseV1: %v", err)
	}
	if sessionResponse.Sequence != 0 || sessionResponse.Epoch != 1 {
		t.Fatalf("session response = %+v", sessionResponse)
	}

	heartbeatRequest, err := ParseHeartbeatRequestV1(readProtocolFixture(t, "heartbeat-request.canonical.txt"))
	if err != nil {
		t.Fatalf("ParseHeartbeatRequestV1: %v", err)
	}
	if heartbeatRequest.Snapshot.Capacity.Available != 1 || heartbeatRequest.Sequence != 1 {
		t.Fatalf("heartbeat request = %+v", heartbeatRequest)
	}

	heartbeatResponse, err := ParseHeartbeatResponseV1(readProtocolFixture(t, "heartbeat-response-lease.canonical.txt"))
	if err != nil {
		t.Fatalf("ParseHeartbeatResponseV1 lease: %v", err)
	}
	if heartbeatResponse.Lease == nil || heartbeatResponse.NoLeaseReason != nil {
		t.Fatalf("heartbeat lease response = %+v", heartbeatResponse)
	}

	noLease, err := ParseHeartbeatResponseV1(readProtocolFixture(t, "heartbeat-response-no-lease.canonical.txt"))
	if err != nil {
		t.Fatalf("ParseHeartbeatResponseV1 no lease: %v", err)
	}
	if noLease.Lease != nil || noLease.NoLeaseReason == nil || *noLease.NoLeaseReason != NoLeaseHostedHold {
		t.Fatalf("heartbeat no-lease response = %+v", noLease)
	}
}

func TestFrozenProtocolDocumentsCanonicalizeIdentically(t *testing.T) {
	for _, name := range []string{
		"session-request",
		"session-response",
		"heartbeat-request",
		"heartbeat-response-lease",
		"heartbeat-response-no-lease",
	} {
		t.Run(name, func(t *testing.T) {
			var document any
			if err := json.Unmarshal(readProtocolFixture(t, name+".json"), &document); err != nil {
				t.Fatalf("unmarshal: %v", err)
			}
			got, err := CanonicalJSON(document)
			if err != nil {
				t.Fatalf("CanonicalJSON: %v", err)
			}
			want := readProtocolFixture(t, name+".canonical.txt")
			if string(got) != string(want) {
				t.Fatalf("canonical = %s, want %s", got, want)
			}
			if _, err := ParseCanonicalJSON(want); err != nil {
				t.Fatalf("ParseCanonicalJSON: %v", err)
			}
		})
	}
}

func TestProtocolParsersRejectMissingUnknownAndInconsistentFields(t *testing.T) {
	missing := mutateProtocolFixture(t, "heartbeat-request.canonical.txt", func(request map[string]any) {
		snapshot := request["snapshot"].(map[string]any)
		delete(snapshot["capacity"].(map[string]any), "queued")
	})
	if _, err := ParseHeartbeatRequestV1(missing); err == nil {
		t.Fatal("heartbeat request accepted a missing nested field")
	}

	unknown := mutateProtocolFixture(t, "heartbeat-response-lease.canonical.txt", func(response map[string]any) {
		response["maintenance"].(map[string]any)["unknown"] = true
	})
	if _, err := ParseHeartbeatResponseV1(unknown); err == nil {
		t.Fatal("heartbeat response accepted an unknown nested field")
	}

	both := mutateProtocolFixture(t, "heartbeat-response-lease.canonical.txt", func(response map[string]any) {
		response["noLeaseReason"] = "hosted-hold"
	})
	if _, err := ParseHeartbeatResponseV1(both); err == nil {
		t.Fatal("heartbeat response accepted both a lease and a no-lease reason")
	}
}

func TestHeartbeatResponseRequiresCheckedExpiryEquation(t *testing.T) {
	for _, test := range []struct {
		name       string
		durationMs float64
		expiry     string
	}{
		{name: "shorter", durationMs: 20000, expiry: "2026-01-01T00:00:20.999Z"},
		{name: "longer", durationMs: 20000, expiry: "2026-01-01T00:00:21.001Z"},
		{name: "duration overflow", durationMs: 9223372036855, expiry: "2026-01-01T00:00:21.000Z"},
	} {
		t.Run(test.name, func(t *testing.T) {
			raw := mutateProtocolFixture(t, "heartbeat-response-lease.canonical.txt", func(response map[string]any) {
				lease := response["lease"].(map[string]any)
				lease["durationMs"] = test.durationMs
				lease["expiry"] = test.expiry
			})
			if _, err := ParseHeartbeatResponseV1(raw); err == nil {
				t.Fatal("heartbeat response accepted an invalid expiry equation")
			}
		})
	}
}

func TestHeartbeatResponseRequiresRoutingHolderModeMatrix(t *testing.T) {
	for _, test := range []struct {
		name   string
		mutate func(map[string]any, map[string]any)
	}{
		{name: "hosted lease", mutate: func(response, _ map[string]any) { response["routingState"] = "HOSTED" }},
		{name: "draining lease", mutate: func(response, _ map[string]any) { response["routingState"] = "DRAINING_TO_HOSTED" }},
		{name: "full state canary mode", mutate: func(response, lease map[string]any) {
			response["routingState"] = "PORTABLE_CANARY"
			lease["mode"] = "enabled"
		}},
		{name: "wrong holder", mutate: func(_ map[string]any, lease map[string]any) { lease["holder"] = "legacy" }},
		{name: "full state canary set", mutate: func(_ map[string]any, lease map[string]any) { lease["canaryScaleSet"] = "canary-set" }},
		{name: "legacy state portable holder", mutate: func(response, _ map[string]any) { response["routingState"] = "LEGACY" }},
	} {
		t.Run(test.name, func(t *testing.T) {
			raw := mutateProtocolFixture(t, "heartbeat-response-lease.canonical.txt", func(response map[string]any) {
				test.mutate(response, response["lease"].(map[string]any))
			})
			if _, err := ParseHeartbeatResponseV1(raw); err == nil {
				t.Fatal("heartbeat response accepted an inconsistent routing matrix")
			}
		})
	}
}

func TestHeartbeatResponseRequiresPositiveLeaseGeneration(t *testing.T) {
	for _, fixture := range []string{
		"heartbeat-response-lease.canonical.txt",
		"heartbeat-response-no-lease.canonical.txt",
	} {
		t.Run(fixture, func(t *testing.T) {
			raw := mutateProtocolFixture(t, fixture, func(response map[string]any) {
				response["maintenance"].(map[string]any)["leaseGeneration"] = float64(0)
				if lease, ok := response["lease"].(map[string]any); ok {
					lease["leaseGeneration"] = float64(0)
				}
			})
			if _, err := ParseHeartbeatResponseV1(raw); err == nil {
				t.Fatal("heartbeat response accepted lease generation zero")
			}
		})
	}
}

func TestProtocolTimestampFormattingIsExactUTCAndMillisecondBounded(t *testing.T) {
	input := time.Date(2026, 1, 1, 1, 2, 3, 456789000, time.FixedZone("offset", 3600))
	if got := FormatProtocolTimestamp(input); got != "2026-01-01T00:02:03.456Z" {
		t.Fatalf("FormatProtocolTimestamp = %q", got)
	}
}

func TestSessionResponseSafeIntegerBoundary(t *testing.T) {
	atBoundary := mutateProtocolFixture(t, "session-response.canonical.txt", func(response map[string]any) {
		response["epoch"] = float64(9007199254740991)
	})
	if _, err := ParseSessionResponseV1(atBoundary); err != nil {
		t.Fatalf("safe boundary rejected: %v", err)
	}
	beyond := mutateProtocolFixture(t, "session-response.canonical.txt", func(response map[string]any) {
		response["epoch"] = float64(9007199254740992)
	})
	if _, err := ParseSessionResponseV1(beyond); err == nil {
		t.Fatal("unsafe integer accepted")
	}
}

func TestProtocolParsersRejectNullForNonNullableWireFields(t *testing.T) {
	t.Run("session scalar", func(t *testing.T) {
		raw := mutateProtocolFixture(t, "session-response.canonical.txt", func(response map[string]any) {
			response["sequence"] = nil
		})
		if _, err := ParseSessionResponseV1(raw); err == nil {
			t.Fatal("session response accepted sequence:null")
		}
	})

	t.Run("heartbeat boolean and counter", func(t *testing.T) {
		for _, field := range []string{"degraded", "runningJobs"} {
			raw := mutateProtocolFixture(t, "heartbeat-request.canonical.txt", func(request map[string]any) {
				request["snapshot"].(map[string]any)[field] = nil
			})
			if _, err := ParseHeartbeatRequestV1(raw); err == nil {
				t.Fatalf("heartbeat request accepted %s:null", field)
			}
		}
	})

	t.Run("lease alias set", func(t *testing.T) {
		raw := mutateLeaseFixture(t, func(lease map[string]any) {
			lease["archivedDisabledAliases"] = nil
		})
		if _, err := ParseLeaseV1(raw); err == nil {
			t.Fatal("lease accepted archivedDisabledAliases:null")
		}
	})
}
